### PR TITLE
feat: add `PERCENTILE_DISC` reduce case operator engine implementation

### PR DESCRIPTION
**Pull request by fucangfy**
1. as the flow of `percentile_disc` is pretty different with other reduce operators, I refactored the code organization, make  `SecretReduceImpl` and `ExecuteInPlain` two methods abstract
2. the logic is derived from the implementation of `group_agg` and `oblivious_group_agg`

### CODE REVIEW

File: engine/operator/reduce.h [hunk 1]
```diff
@@ -134,4 +141,35 @@ class ReduceMax : public ReduceBase {
...
 
+class ReducePercentileDisc : public ReduceBase {
+ public:
+  static const std::string kOpType;
+  static constexpr char kPercent[] = "percent";
+  const std::string& Type() const override;
+
+ protected:
+  std::string GetArrowFunName() override {
+    YACL_THROW("unsupported reduce func: percentile_disc");
```
Comment (tongke6) [inline engine/operator/reduce.h:152]: looks like quantile in arrow? https://arrow.apache.org/docs/python/generated/pyarrow.compute.quantile.html

Comment (fucangfy) [reply engine/operator/reduce.h:152]: yeah, looks like a more appropriate function, let me try with this. thanks a lot.

Comment (fucangfy) [reply engine/operator/reduce.h]: it's done with `group_agg` as well, it looks more elegant, really appreciate for the information

**Comment by fucangfy** [comment]
add some more test cases for all percentile_disc scenarios, maybe some have duplicate code in the testing cases, I will find some time to refactor this.

**Review by tongke6 — APPROVED** [review]
LGTM

File: engine/operator/reduce.cc [hunk 2]
```diff
@@ -222,4 +230,104 @@ ReduceBase::ReduceFn ReduceMax::GetReduceFn(spu::SPUContext* sctx) {
...
+               "percent should be in [0, 1], but got={}", percent_);
+}
+
+void ReducePercentileDisc::AggregateInit(spu::SPUContext* sctx,
+                                         const spu::Value& in) {
+  // do nothing
+}
+
+spu::Value ReducePercentileDisc::GetInitValue(spu::SPUContext* sctx) {
+  YACL_THROW("unsupported reduce func: percentile_disc");
```
Comment (tongke6) [inline engine/operator/reduce.cc:252]: nit: exception message is suggested to be changed to something like "should not reach here"

File: engine/operator/reduce.cc [hunk 3]
```diff
@@ -222,4 +230,104 @@ ReduceBase::ReduceFn ReduceMax::GetReduceFn(spu::SPUContext* sctx) {
...
+}
+
+ReduceBase::ReduceFn ReducePercentileDisc::GetReduceFn(spu::SPUContext* sctx) {
+  YACL_THROW("unsupported reduce func: percentile_disc");
```
Comment (tongke6) [inline engine/operator/reduce.cc:256]: ditto

**Review by tongke6 — COMMENTED** [review]
LGTM

File: engine/operator/reduce.cc (hunk 2 above)
Comment (fucangfy) [reply engine/operator/reduce.cc]: done

File: engine/operator/reduce.cc (hunk 3 above)
Comment (fucangfy) [reply engine/operator/reduce.cc]: done

**Review by tongke6 — APPROVED** [review]
LGTM
